src/Domain/DomainObject.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace Baikal\Domain;
+
+abstract class DomainObject {
+
+    /**
+     * Creates an instance of the Domain object, based on an array of values.
+     *
+     * This method will throw an InvalidArgumentException if unknown values
+     * are passed.
+     */
+    static function fromArray(array $values) {
+
+        $obj = new static();
+        foreach ($values as $k => $v) {
+            if (property_exists($obj, $k)) {
+                $obj->$k = $v;
+            } else {
+                throw new \InvalidArgumentException('Invalid property: ' . $k . ' for class ' . get_class($obj));
+            }
+        }
+        return $obj;
+
+    }
+
+    /**
+     * Creates an instance of the Domain object, based on an array of values.
+     *
+     * This function works identical to fromArray, but unknown keys are
+     * ignored.
+     */
+    static function fromPostForm(array $values) {
+
+        $obj = new static();
+        foreach ($values as $k => $v) {
+            if (property_exists($obj, $k)) {
+                $obj->$k = $v;
+            }
+        }
+        return $obj;
+
+    }
+
+    /**
+     * Validator
+     *
+     * Called before insert and update operations
+     *
+     * Should throw an \InvalidArgumentException if there's a validation problem.
+     * @return void
+     */
+    function validate() {
```

Comment:
The issue I have is that what's valid for creation, is not necessarily valid for updating. This is also why I added two different functions. A solution to that is to make more of a CQRS-style system where we're not modelling a User, but rather a UserCreationOperation and UserUpdateOperation, but I'm also a bit worried there about creating too much complexity.

I agree that the 'opt-in' can be a bit dangerous, and it's also not amazing that a Domain object can have a valid or invalid state, but I thought this approach would make it the simplest for a intermediate dev to fully grasp where what is being done... 

I'm open to suggestions though. This is simply what I came up with looking at your existing source, and having to make it work predictably and correctly with the UserRepository (which was incomplete before).